File: checks/ruby/lib/section508/checks/heading_elements.rb
# frozen_string_literal: true

module Section508
  module Checks
    # Section names are heading elements, not a bold paragraph.
    class HeadingElements
      def self.call(root)
        return [] unless root

        fakes = fake_headings(root)
        return [] if fakes.empty?

        sample = Dom.norm(fakes.first)
        if sample.empty?
          ["A section name is a bold paragraph, not a heading element."]
        else
          ["“#{sample}” is a bold paragraph, not a heading element."]
        end
      end

      def self.fake_headings(root)
        return [] unless root.respond_to?(:css)

        Dom.elements(root, "p, div").select do |element|
          next false if element.at_css("p, div, ul, ol, table, h1, h2, h3, h4, h5, h6")

          significant = element.children.reject { |node| node.text? && node.text.strip.empty? }
          next false unless significant.length == 1

          child = significant.first
          child.element? && %w[STRONG B].include?(child.name.upcase)
        end
      end
    end
  end
end

File: checks/python/src/section508/checks/heading_elements.py
from section508.dom import is_tag, norm_text, select, significant_children, tag_name


def check(root):
    if not root:
        return []

    fakes = fake_headings(root)
    if not fakes:
        return []

    sample = norm_text(fakes[0])
    if sample:
        return [f"“{sample}” is a bold paragraph, not a heading element."]
    return ["A section name is a bold paragraph, not a heading element."]


def fake_headings(root):
    fakes = []
    for element in select(root, "p, div"):
        if select(element, "p, div, ul, ol, table, h1, h2, h3, h4, h5, h6"):
            continue
        significant = significant_children(element)
        if len(significant) != 1:
            continue
        child = significant[0]
        if is_tag(child) and tag_name(child) in {"strong", "b"}:
            fakes.append(element)
    return fakes
